### PR TITLE
[TT-Train] Add Comprehensive GELU Activation Test Suite #34824

### LINKED ISSUE #34824: [TT-Train] Subtask: Add Comprehensive GELU Activation Test Suite
# Subtask: Add Comprehensive GELU Activation Test Suite

**Parent Issue:** [Implement BERT model in TTML and benchmark against PyTorch and TTNN implementations #26739](https://github.com/tenstorrent/tt-metal/issues/26739)

---

## Title

Add comprehensive GELU activation test suite for BERT implementation

---

## Description

The GELU (Gaussian Error Linear Unit) activation function is a critical component of BERT and other transformer models. This subtask adds comprehensive test coverage for the GELU operation to validate correctness before BERT model integration.

### Scope

Add a production-ready test suite with **26 tests** covering:

1. **Basic Correctness (8 tests)**
   - Forward/backward passes validation
   - Deterministic values with known input/output pairs
   - Block alignment patterns (Wt % 4 = 0,1,2,3) for tile-based hardware
   - Large tensors (up to 1M elements) for memory stress testing

2. **BERT Integration (7 tests)**
   - BERT-base shapes: hidden_dim=768, intermediate_dim=3072
   - BERT-large shapes: hidden_dim=1024, intermediate_dim=4096
   - Maximum sequence length (512 tokens)
   - MLP integration test simulating `Linear → GELU → Linear` pattern

3. **Numerical Stability (6 tests)**
   - Saturation behavior at extreme values
   - Near-zero behavior validation
   - Gradient flow and accumulation tests

4. **Edge Cases (3 tests)**
   - NaN/Inf propagation through BFloat16 pipeline
   - Extreme saturation (±50, ±100)
   - Minimal tensor test case

5. **Memory Configurations (2 tests)**
   - L1 and DRAM memory configs (parameterized)

### Implementation Details

- Reference implementations for exact GELU: `GELU(x) = 0.5 * x * (1 + erf(x/sqrt(2)))`
- BFloat16-appropriate tolerances (rtol=1e-3, atol=3e-2)
- Deterministic seeds for reproducible test results
- Comprehensive documentation with mathematical properties

### Files Changed

| File | Change |
|------|--------|
| `tt-train/tests/ops/gelu_op_test.cpp` | +607 lines (new file) |
| `tt-train/tests/CMakeLists.txt` | +1 line (add to build) |

### Acceptance Criteria

- [ ] All 26 tests pass on Wormhole hardware
- [ ] Tests cover BERT-base and BERT-large configurations
- [ ] Edge cases (NaN, Inf, extreme values) are documented
- [ ] Deterministic seeds ensure reproducibility
- [ ] No new dependencies added

---

## Labels

- `tt-train`
- `testing`
- `BERT`

### PR DESCRIPTION
**Pull request by epam-iaroslav-voitovych**
### Ticket

Closes #34824

Related: [Implement BERT model in TTML #26739](https://github.com/tenstorrent/tt-metal/issues/26739)

### Problem description

The GELU (Gaussian Error Linear Unit) activation function is a critical component of BERT and other transformer models. Before this change, tt-train lacked comprehensive test coverage for the GELU operation, making it difficult to:

1. Validate correctness of the exact erf-based GELU implementation: `GELU(x) = 0.5 * x * (1 + erf(x/sqrt(2)))`
2. Ensure numerical stability across the full input range
3. Verify hardware behavior for edge cases (NaN, Inf, extreme values)
4. Confirm correct operation with BERT-specific tensor shapes
5. Test both L1 and DRAM memory configurations

### What's changed

Added a production-ready test suite with **26 tests** covering all aspects of GELU activation.

**Test Coverage:**

| Category | Tests | Description |
|----------|-------|-------------|
| Basic Correctness | 8 | Forward/backward passes, deterministic values, block alignment patterns | | BERT Integration | 7 | BERT-base (768/3072), BERT-large (1024/4096), max sequence length, MLP integration | | Numerical Stability | 6 | Saturation, near-zero behavior, gradient flow, gradient accumulation, precision | | Edge Cases | 3 | NaN/Inf propagation, extreme saturation (±50, ±100), minimal tensor | | Memory Configurations | 2 | L1 and DRAM memory configs (parameterized test) |

**Implementation Details:**

- Reference implementations for exact GELU forward and backward passes
- BFloat16-appropriate tolerances (rtol=1e-3, atol=3e-2)
- Proper TTML conventions using `core::parallel_generate` for random data
- Deterministic seeds for reproducible test results
- Comprehensive documentation with mathematical properties

**Key Test Cases:**

1. **Block Alignment** - Tests Wt % 4 = 0,1,2,3 patterns for tile-based hardware
2. **BERT MLP Integration** - Simulates `Linear → GELU → Linear` pattern used in transformer FFN
3. **NaN/Inf Propagation** - Documents observed hardware behavior:
   - NaN → +Inf
   - +Inf → +Inf
   - -Inf → 0
4. **Large Tensors** - Up to 1M elements for memory stress testing

**Files Changed:**

| File | Change |
|------|--------|
| `tt-train/tests/ops/gelu_op_test.cpp` | +607 lines (new file) | | `tt-train/tests/CMakeLists.txt` | +1 line (add to build) |

**Test Results:** All 26 tests pass on Wormhole hardware.

**NIGHTLY Tests:**
Two tests are marked NIGHTLY for longer CI runs:
- `NIGHTLY_GELU_VeryLarge` - 1M elements stress test
- `NIGHTLY_GELU_LargeRandom` - Large random data validation

**Backward Compatibility:** This is purely additive - no breaking changes to existing code or tests.

**Dependencies:** No new dependencies added.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by ayerofieiev-tt — COMMENTED** [review]
@vmelnykovTT ptal

**Comment by ayerofieiev-tt** [comment]
One think I am not sure if good is - how we do the accuracy check. I think we ideally should do both all_close and ulp check.
Would be good to pick @nmauriceTT opinion.

**Comment by nmauriceTT** [comment]
@ayerofieiev-tt Doing both would do no harm. Although, if the ULP check then the allclose check will probably pass.

When dealing with a specified data type, and when aiming for high-accuracy, ULP is easier to reason about because it tells you by 'how many consecutive floating point numbers' you are off. It's also a good way to know if you are dealing with rounding errors or not. 

allclose is good/convenient if you don't mind large relative errors near 0 (as long as absolute error is below the specified threshold). On TT hardware, that also mean not having to worry about subnormals (absolute threshold will likely be larger than subnormals). 


If I were to comment about something in this PR: if you are going to generate a tensor with 10'000 random elements and test a bfloat16 implementation, then you may as well just generate all possible bfloat16 values (65'536 =33,426 = 256 x 256 tensor) and test on the entire useful range (just set special values and subnormals to 0). That's something we have been doing with (e.g. with [ttnn.exp tests](https://github.com/tenstorrent/tt-metal/blob/246986756ea5af7799b33ab9528ad25c2e343f2f/tests/ttnn/unit_tests/operations/eltwise/test_exp_21f.py#L17)
In C++, this should be a matter of allocating an integer tensor, initializing it with `std::iota` or equivalent, and then reinterpreting as float/bfloat16.

**Comment by ivoitovych** [comment]
@ayerofieiev-tt @nmauriceTT

Thanks for the detailed review and great suggestions on ULP checks and exhaustive bfloat16 coverage!

I took your feedback and started implementing both:
- Added ULP-based validation alongside the existing checks.
- Extended coverage to better stress the implementation (including more targeted patterns in problematic regions).

While doing this, the tighter checks revealed significant accuracy issues in the current `ttnn.gelu()` accurate mode implementation – catastrophic ULP errors (up to 32,767) in deep negative tails, near-zero inputs, and the transition region around x ≈ -5.

I performed a deeper investigation:
- Amended test suite with ULP diagnostics: https://github.com/tenstorrent/tt-metal/compare/main...ivoitovych:tt-metal:ivoitovych/bert-model-for-ttml-pr-gelu-test-suite-amendment
- Dedicated diagnostic branch showing the spikes: https://github.com/tenstorrent/tt-metal/compare/main...ivoitovych:tt-metal:ivoitovych/bert-model-for-ttml-pr-gelu-test-suite-amendment-ulp-diagnostic-04

GELU UPL Error Spikes

Separate research repo evaluating 39 different GELU approximations exhaustively over all 65,536 bfloat16 values: https://github.com/ivoitovych/bf16_gelu_research  
Key finding: max ULP ≤ 1 is achievable with adaptive polynomials or pure arithmetic methods, and several options achieve ≤ 35 max ULP.

I reported the underlying accuracy bug here: https://github.com/tenstorrent/tt-metal/issues/35290  
The issue includes detailed reproducers, root-cause analysis, and concrete fix proposals.

In the meantime, to avoid noisy failures in this test suite due to the known bug, I can mark the most affected tests as disabled (or expected-to-fail) until the fix lands. Let me know if you'd like me to push that change.

Thanks!

File: tt-train/tests/ops/gelu_op_test.cpp [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+
+TEST_F(GELUOpTest, GELU_BERT_MultiBatch) {
+    // Training scenario with larger batch
+    CompareGELUVsReferenceWithShape({4, 1, 128, 768});
+}
+
+TEST_F(GELUOpTest, GELU_BERTMLPIntegration) {
+    using namespace ttml;
+
+    // Simulate BERT MLP pattern: Linear → GELU → Linear
```
Comment (aliaksei-sala) [inline tt-train/tests/ops/gelu_op_test.cpp:296]: The test assumes there are two linear layers, but I don’t see them. 
Am I missing something?

**Comment by mdragulaTT** [comment]
Hi @epam-iaroslav-voitovych, I really appreciate you following the established test style in tt-train! The consistency in testing approach is valuable for maintainability.

And I'm so sorry to reply so late. I know you have opened the PR a long time ago, but I didn't have time to look at this earlier.

However, I'm wondering if testing GELU in tt-train is the right approach. From my understanding, tt-train tests are primarily meant to validate kernels that we've internally fused within tt-train (like SiLU, RMSNorm, etc.). Since GELU isn't currently fused by our team, it seems like this would be better tested in the ttnn test suite.

### Fusion in tt-train

The way I understand it, there are some automated methods to fuse operations, but they're sometimes suboptimal. That's why we implement our own custom fused kernels in tt-train like:
- RMSNorm (forward/backward) [`operations.hpp:22-26`](https://github.com/tenstorrent/tt-metal/blob/main/tt-train/sources/ttml/metal/operations.hpp#L22-L26)
- SiLU (backward) [`operations.hpp:48-49`](https://github.com/tenstorrent/tt-metal/blob/main/tt-train/sources/ttml/metal/operations.hpp#L48-L49)
- CrossEntropy (forward/backward) [`operations.hpp:34-40`](https://github.com/tenstorrent/tt-metal/blob/main/tt-train/sources/ttml/metal/operations.hpp#L34-L40)
- SGD fused optimizer [`operations.hpp:57-58`](https://github.com/tenstorrent/tt-metal/blob/main/tt-train/sources/ttml/metal/operations.hpp#L57-L58)

These are custom implementations that we maintain in tt-train for performance reasons. But GELU, as far as I know, is not implemented as a custom fused kernel in tt-train - it's just using the standard ttnn operation.

### Existing ttnn Test Coverage

There are already several GELU tests that exist in the ttnn test suite. I have found the following tests with our LLM:

#### Unit Tests:
- [`test_unary_gelu_ttnn`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py#L392-L402) - Basic functionality with different shapes and approximation modes
- [`test_gelu_accurate_allclose`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/unit_tests/operations/eltwise/test_activation.py#L90-L110) - Accuracy testing across different input ranges
- [`test_gelu`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/unit_tests/operations/eltwise/test_unary.py#L217-L218) - Simple functionality test

#### Sweep Tests:
- Framework sweep testing various shapes, data types, and memory configurations [`gelu.py:26-73`](https://github.com/tenstorrent/tt-metal/tree/main/tests/sweep_framework/sweeps/eltwise/unary/gelu/gelu.py#L26-L73)
- TTNN sweep tests for different tensor dimensions [`gelu.py:26-52`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/sweep_tests/sweeps/sweeps/gelu.py#L26-L52)

#### Backward Pass Tests:
- Standard backward pass with different approximation modes [`test_backward_gelu.py:26-36`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py#L26-L36)
- Fused backward operations [`test_backward_gelu_fused.py:47-57`](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu_fused.py#L47-L57)

### Testing Philosophy

I'm not a ttnn specialist and don't know the full history of how the tests were developed and monitored, but my understanding is that if someone is using a standard ttnn operation, it should be tested in ttnn rather than tt-train. I'm not a big fan of keeping tests for non-tt-train operations in multiple places - it leads to duplication and maintenance overhead.

If we do end up fusing GELU in the future (implementing a custom kernel in tt-train), then absolutely creating tt-train tests would make sense! But for now, since what you want to test is actually not implemented in tt-train (if I'm not mistaken), perhaps we could focus on enhancing the ttnn test coverage if needed.

What do you think?
